### PR TITLE
fix: give up after some specified attempts of repainting the prompt

### PR DESCRIPTION
**Pull request by lavafroth**
## Description

- Adds a repaint attempts counter to `ReaderData` struct
- The number of retries can bet set using a newly introduced `fish_redraw_attempts` environment variable
- The default value is set to 3
- If the subshell exit status of the prompt command is not zero for `threshold` number of times,
`redraw_attempt.exceeded()` is true and we paint the default prompt

Fixes issue #9796

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

**Comment by faho** [comment]
> The number of retries can bet set using a newly introduced fish_redraw_attempts environment variable

Just as a general thing: I wouldn't introduce a variable like that unless we can justify why someone would want to set it.

We try not to introduce options, if we can.

Just hard-code it at 3.

------------

Also I am reasonably sure this will end up calling `fish_job_summary` 3 times, and print the message 3 times, in the case of #9796.

Which is definitely better than looping endlessly, but of course still not *ideal*. I don't know how to achieve that ideal, tho.

File: src/reader.rs [hunk 1]
```diff
@@ -922,6 +942,24 @@ pub fn reader_set_autosuggestion_enabled(vars: &dyn Environment) {
...
 }
 
+/// Change the number of attempts to redraw a prompt before giving up.
+pub fn reader_set_handle_redraw_attempts_change(vars: &dyn Environment) {
+    // We don't need to _change_ if we're not initialized yet.
+    let Some(data) = current_data() else {
+        return;
+    };
+    let attempts: usize = vars
+        .get(L!("fish_redraw_attempts"))
```
Comment (faho) [inline src/reader.rs:952]: Remove this variable and just hard-code the number of attempts.

**Comment by lavafroth** [comment]
I'll get back to this by tonight.

File: src/reader.rs [hunk 2]
```diff
@@ -4168,16 +4210,20 @@ impl<'a> Reader<'a> {
...
+                    if zelf.redraw_attempt.exceeded() || left_prompt_deleted {
+                        zelf.redraw_attempt.reset();
                         DEFAULT_PROMPT
                     } else {
                         &zelf.conf.left_prompt_cmd
                     },
                     zelf.parser,
                     Some(&mut prompt_list),
-                    /*apply_exit_status=*/ false,
+                    /*apply_exit_status=*/ true,
```
[the remark this answers is not in the document]
Comment (lavafroth) [reply src/reader.rs]: Turns out `exec_subshell` returns `$status` of the command executed anyways. Therefore, we would not have to apply the exit status.

File: src/reader.rs [hunk 3]
```diff
@@ -543,6 +543,26 @@ pub struct ReaderData {
     in_flight_autosuggest_request: WString,
 
     rls: Option<ReadlineLoopState>,
+    redraw_attempt: RedrawAttempt,
+}
+
+pub struct RedrawAttempt {
+    threshold: usize,
```
Comment (krobelus) [inline src/reader.rs:550]: if we stick with this approach: this field is unused and should be removed (which unlocks further simplification)

**Comment by lavafroth** [comment]
Closing this in favor of #10891

**Pull request closed without merging**